File: apps/campus_hub_app/lib/features/portfolio/data/portfolio_repository.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../../core/network/api_client.dart';
import '../domain/portfolio_models.dart';

class PortfolioRepository {
  final Dio _dio;

  PortfolioRepository(this._dio);

  Future<PortfolioModel> getUserPortfolio() async {
    final response = await _dio.get('/api/v1/portfolio/me');
    return PortfolioModel.fromJson(response.data['data']);
  }

  Future<PortfolioModel> getPublicPortfolio(String identifier) async {
    final response = await _dio.get('/api/v1/portfolio/public/$identifier');
    return PortfolioModel.fromJson(response.data['data']);
  }

  Future<PortfolioModel> updatePortfolio({
    String? bio,
    String? githubUrl,
    String? linkedinUrl,
    String? websiteUrl,
    String? resumeUrl,
    double? cgpa,
    String? customUsername,
    bool? isPublic,
  }) async {
    final response = await _dio.patch(
      '/api/v1/portfolio/me',
      data: {
        'bio': bio,
        'github_url': githubUrl,
        'linkedin_url': linkedinUrl,
        'website_url': websiteUrl,
        'resume_url': resumeUrl,
        'cgpa': cgpa,
        'custom_username': customUsername,
        'is_public': isPublic,
      },
    );
    return PortfolioModel.fromJson(response.data['data']);
  }

  Future<void> addProject({
    required String title,
    String? description,
    List<String>? techStack,
    String? projectUrl,
    String? repoUrl,
    String? imageUrl,
  }) async {
    await _dio.post(
      '/api/v1/portfolio/projects',
      data: {
        'title': title,
        'description': description,
        'tech_stack': techStack,
        'project_url': projectUrl,
        'repo_url': repoUrl,
        'image_url': imageUrl,
      },
    );
  }

  Future<void> deleteProject(String projectId) async {
    await _dio.delete('/api/v1/portfolio/projects/$projectId');
  }

  Future<void> addSkill({
    required String skillName,
    String? category,
    String? proficiency,
  }) async {
    await _dio.post(
      '/api/v1/portfolio/skills',
      data: {
        'skill_name': skillName,
        'category': category,
        'proficiency': proficiency,
      },
    );
  }

  Future<void> deleteSkill(String skillId) async {
    await _dio.delete('/api/v1/portfolio/skills/$skillId');
  }

  Future<void> addCertificate({
    required String title,
    required String issuer,
    String? issueDate,
    String? credentialUrl,
    String? credentialId,
  }) async {
    await _dio.post(
      '/api/v1/portfolio/certificates',
      data: {
        'title': title,
        'issuer': issuer,
        'issue_date': issueDate,
        'credential_url': credentialUrl,
        'credential_id': credentialId,
      },
    );
  }

  Future<void> deleteCertificate(String certId) async {
    await _dio.delete('/api/v1/portfolio/certificates/$certId');
  }

  Future<void> addAchievement({
    required String title,
    String? category,
    String? description,
    String? dateAchieved,
    String? proofUrl,
  }) async {
    await _dio.post(
      '/api/v1/portfolio/achievements',
      data: {
        'title': title,
        'category': category,
        'description': description,
        'date_achieved': dateAchieved,
        'proof_url': proofUrl,
      },
    );
  }

  Future<void> deleteAchievement(String achievementId) async {
    await _dio.delete('/api/v1/portfolio/achievements/$achievementId');
  }
}

final portfolioRepositoryProvider = Provider<PortfolioRepository>((ref) {
  final dio = ref.watch(dioClientProvider);
  return PortfolioRepository(dio);
});

File: backend/src/modules/portfolio/portfolio.repository.ts
import { prisma } from '../../config/database';
import {
  UpdatePortfolioDto,
  AddProjectDto,
  AddSkillDto,
  AddCertificateDto,
  AddAchievementDto,
} from './portfolio.types';

export class PortfolioRepository {
  async findPortfolioByUserId(userId: string) {
    let portfolio = await prisma.portfolio.findUnique({
      where: { user_id: userId },
      include: {
        user: {
          select: {
            id: true,
            first_name: true,
            last_name: true,
            email: true,
            avatar_url: true,
            role: true,
            department: { select: { id: true, name: true, code: true } },
          },
        },
        projects: { orderBy: { created_at: 'desc' } },
        skills: true,
        certificates: { orderBy: { created_at: 'desc' } },
        achievements: { orderBy: { created_at: 'desc' } },
      },
    });

    if (!portfolio) {
      portfolio = await prisma.portfolio.create({
        data: { user_id: userId },
        include: {
          user: {
            select: {
              id: true,
              first_name: true,
              last_name: true,
              email: true,
              avatar_url: true,
              role: true,
              department: { select: { id: true, name: true, code: true } },
            },
          },
          projects: true,
          skills: true,
          certificates: true,
          achievements: true,
        },
      });
    }

    return portfolio;
  }

  async findPublicPortfolio(identifier: string) {
    return prisma.portfolio.findFirst({
      where: {
        is_public: true,
        OR: [{ user_id: identifier }, { custom_username: identifier }],
      },
      include: {
        user: {
          select: {
            id: true,
            first_name: true,
            last_name: true,
            email: true,
            avatar_url: true,
            role: true,
            department: { select: { id: true, name: true, code: true } },
          },
        },
        projects: { orderBy: { created_at: 'desc' } },
        skills: true,
        certificates: { orderBy: { created_at: 'desc' } },
        achievements: { orderBy: { created_at: 'desc' } },
      },
    });
  }

  async updatePortfolio(userId: string, dto: UpdatePortfolioDto) {
    const portfolio = await this.findPortfolioByUserId(userId);

    return prisma.portfolio.update({
      where: { id: portfolio.id },
      data: {
        bio: dto.bio,
        github_url: dto.github_url,
        linkedin_url: dto.linkedin_url,
        website_url: dto.website_url,
        resume_url: dto.resume_url,
        cgpa: dto.cgpa,
        custom_username: dto.custom_username,
        is_public: dto.is_public,
      },
      include: {
        projects: true,
        skills: true,
        certificates: true,
        achievements: true,
      },
    });
  }

  async addProject(userId: string, dto: AddProjectDto) {
    const portfolio = await this.findPortfolioByUserId(userId);

    return prisma.portfolioProject.create({
      data: {
        portfolio_id: portfolio.id,
        title: dto.title,
        description: dto.description,
        tech_stack: dto.tech_stack || [],
        project_url: dto.project_url,
        repo_url: dto.repo_url,
        image_url: dto.image_url,
      },
    });
  }

  async deleteProject(userId: string, projectId: string) {
    const portfolio = await this.findPortfolioByUserId(userId);
    return prisma.portfolioProject.deleteMany({
      where: { id: projectId, portfolio_id: portfolio.id },
    });
  }

  async addSkill(userId: string, dto: AddSkillDto) {
    const portfolio = await this.findPortfolioByUserId(userId);

    return prisma.portfolioSkill.create({
      data: {
        portfolio_id: portfolio.id,
        skill_name: dto.skill_name,
        category: dto.category || 'General',
        proficiency: dto.proficiency || 'Intermediate',
      },
    });
  }

  async deleteSkill(userId: string, skillId: string) {
    const portfolio = await this.findPortfolioByUserId(userId);
    return prisma.portfolioSkill.deleteMany({
      where: { id: skillId, portfolio_id: portfolio.id },
    });
  }

  async addCertificate(userId: string, dto: AddCertificateDto) {
    const portfolio = await this.findPortfolioByUserId(userId);

    return prisma.portfolioCertificate.create({
      data: {
        portfolio_id: portfolio.id,
        title: dto.title,
        issuer: dto.issuer,
        issue_date: dto.issue_date ? new Date(dto.issue_date) : null,
        credential_url: dto.credential_url,
        credential_id: dto.credential_id,
      },
    });
  }

  async deleteCertificate(userId: string, certificateId: string) {
    const portfolio = await this.findPortfolioByUserId(userId);
    return prisma.portfolioCertificate.deleteMany({
      where: { id: certificateId, portfolio_id: portfolio.id },
    });
  }

  async addAchievement(userId: string, dto: AddAchievementDto) {
    const portfolio = await this.findPortfolioByUserId(userId);

    return prisma.portfolioAchievement.create({
      data: {
        portfolio_id: portfolio.id,
        title: dto.title,
        category: dto.category || 'General',
        description: dto.description,
        date_achieved: dto.date_achieved ? new Date(dto.date_achieved) : null,
        proof_url: dto.proof_url,
      },
    });
  }

  async deleteAchievement(userId: string, achievementId: string) {
    const portfolio = await this.findPortfolioByUserId(userId);
    return prisma.portfolioAchievement.deleteMany({
      where: { id: achievementId, portfolio_id: portfolio.id },
    });
  }
}
